### DISCUSSION TITLE
Thread safety

### DISCUSSION THREAD

**Discussion by davidk88** [h2zero/NimBLE-Arduino#629] [Q&A]
Hello,

is this library thread safe? I'm using scanning, server and client at the same time on ESP32-c3. Is it allowed to use it like this (pseudocode):

task1_server:
while(true)
 if data_available:
  notify_client()

task2_client:
while(true)
 connect_to_server()
 write_characteristic()
 read_characteristic()
 disconnect()
 delay(...)

I'm getting heap corruption, but for now I did not pinpoint where or why it happens.

Best regards

**Comment by h2zero** [comment]
Yes, this is thread safe, or at least should be, to the best of my knowledge. Analysis of the backtrace/logs should indicate the location of the corruption you are seeing.
